servant-server/src/Servant/Server/Internal.hs

```diff
@@ -176,95 +180,33 @@ methodRouterHeaders method proxy status action = LeafRouter route'
       | pathIsEmpty request =
           let accH    = fromMaybe ct_wildcard $ lookup hAccept $ requestHeaders request
           in runAction (action `addMethodCheck` methodCheck method request
-                               `addAcceptCheck` acceptCheck proxy  accH
+                               `addAcceptCheck` acceptCheck proxy accH
                        ) respond $ \ output -> do
                 let headers = getHeaders output
                     handleA = handleAcceptH proxy (AcceptHeader accH) (getResponse output)
                 processMethodRouter handleA status method (Just headers) request
       | otherwise = respond $ Fail err404
 
-methodRouterEmpty :: Method
-                  -> Delayed (ExceptT ServantErr IO ())
-                  -> Router
-methodRouterEmpty method action = LeafRouter route'
-  where
-    route' request respond
-      | pathIsEmpty request = do
-          runAction (addMethodCheck action (methodCheck method request)) respond $ \ () ->
-            Route $! responseLBS noContent204 [] ""
-      | otherwise = respond $ Fail err404
-
--- | If you have a 'Delete' endpoint in your API,
--- the handler for this endpoint is meant to delete
--- a resource.
---
--- The code of the handler will, just like
--- for 'Servant.API.Get.Get', 'Servant.API.Post.Post' and
--- 'Servant.API.Put.Put', run in @ExceptT ServantErr IO ()@.
--- The 'Int' represents the status code and the 'String' a message
--- to be returned. You can use 'Control.Monad.Trans.Except.throwE' to
--- painlessly error out if the conditions for a successful deletion
--- are not met.
 instance OVERLAPPABLE_
-         ( AllCTRender ctypes a
-         ) => HasServer (Delete ctypes a) where
-
-  type ServerT (Delete ctypes a) m = m a
-
-  route Proxy = methodRouter methodDelete (Proxy :: Proxy ctypes) ok200
+         ( AllCTRender ctypes a, ReflectMethod method, KnownNat status
+         ) => HasServer (Verb method status ctypes a) where
 
-instance OVERLAPPING_
-         HasServer (Delete ctypes ()) where
-
-  type ServerT (Delete ctypes ()) m = m ()
+  type ServerT (Verb method status ctypes a) m = m a
 
-  route Proxy = methodRouterEmpty methodDelete
+  route Proxy = methodRouter method (Proxy :: Proxy ctypes) status
+    where method = reflectMethod (Proxy :: Proxy method)
+          status = toEnum . fromInteger $ natVal (Proxy :: Proxy status)
```

Comment:
We need a `Network.HTTP.Types.Status`, which doesn't have a `Num` instance (but does have an `Enum` instance).